flaskr_tests.py

```diff
@@ -0,0 +1,56 @@
+import os
+import flaskr
+import unittest
+import tempfile
+
+class FlaskrTestCase(unittest.TestCase):
+
+	def setUp(self):
+		self.db_fd, flaskr.app.config['DATABASE'] = tempfile.mkstemp()
+		flaskr.app.config['TESTING'] = True
+		self.app = flaskr.app.test_client()
+		with flaskr.app.app_context():
+			flaskr.init_db()
+
+	def tearDown(self):
+		os.close(self.db_fd)
+		os.unlink(flaskr.app.config['DATABASE'])
+
+	def test_empty_db(self):
+		rv = self.app.get('/')
+		assert b'No entries here so far' in rv.data
+
+	def login(self, username, password):
+		return self.app.post('/login', data=dict(
+			username=username,
+			password=password
+			), follow_redirects=True)
+
+	def logout(self):
+		return self.app.get('logout', follow_redirects=True)
+
+	def test_login_logout(self):
+		rv = self.login('admin','default')
+		assert b'You were logged in' in rv.data
+		
+		rv = self.logout()
+		assert b'You were logged out' in rv.data
+
+		rv = self.login('adminx','default')
+		assert b'Invalid username' in rv.data
+		
+		rv = self.login('admin','defaultx')
+		assert b'Invalid password' in rv.data
+
+	def test_messages(self):
+		self.login('admin', 'default')
+		rv = self.app.post('/add', data=dict(
+			title='<Hello>',
+			text='<strong>HTML</strong> allowed here'
+			), follow_redirects=True)
+		assert b'No entries here so far' not in rv.data
+		assert b'&lt;Hello&gt;' in rv.data
+		assert b'<strong>HTML</strong> allowed here' in rv.data
```

Comment:
Actually, you have the right feeling here - there is a code duplication because you have similar setup for different tests (you need to have a post to be able then edit or delete it).

The good place for this initialization is the setUp() method. You just move similar tests to the separate test class and move all the preparations into the setUp(), see https://docs.python.org/3/library/unittest.html#organizing-test-code.

Other approaches to this code duplication:

- You also can have various helper methods / functions and move the common code there.

- You can use fixtures, this is the common way to pre-create some objects needed for your tests, I am currently using [pytest](http://doc.pytest.org/en/latest/) and it has very nice fixtures system, which is more generic than the `setUp` approach in unittest module. But for now, to aviod mental overhead, that's fine to use setUp or helper methods / functions.

And in the any case, this is out of scope of the current PR and should be done on the separate branch.